### PR TITLE
Add guidelines on code reviews on the team

### PR DESCRIPTION
**Pull request by JohanLorenzo**
This page encompasses what the team discussed 3 weeks ago. 

Let's do a first pass of review on this. I'll get approval from the team before merging.

### CODE REVIEW

File: explanations/best-practices/code-reviews.rst
Comment (bhearsum) [inline explanations/best-practices/code-reviews.rst:21]: This isn't rendering properly (see https://github.com/mozilla-releng/relengdocs/blob/b1b919c390b4256a11c0ac73158cee20c7e48074/explanations/best-practices/code-reviews.rst).

File: explanations/best-practices/code-reviews.rst
Comment (bhearsum) [inline explanations/best-practices/code-reviews.rst:22]: "when the bug report is require lots of..." -> "when the bug report requires lots of..."

File: explanations/best-practices/code-reviews.rst
Comment (bhearsum) [inline explanations/best-practices/code-reviews.rst:23]: On GitHub, this essentially means submitting everything as a Draft. If that's what we want, I think it should be called out explicitly.

File: explanations/best-practices/code-reviews.rst
Comment (bhearsum) [inline explanations/best-practices/code-reviews.rst:34]: The "tests" part deserves some clarification. It's still important to review the contents of tests to ensure that they're testing what's needed, and that the tests themselves don't have bugs in them.

**Review by bhearsum — CHANGES_REQUESTED** [review]
Thanks for rolling this up! There's a few tweaks needed, I think, but it's largely good.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst]: Good catch. Fixed.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst]: Added. I also added the Phab equivalent.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst]: Added another sentence. Let me know what you think!

File: explanations/best-practices/code-reviews.rst
Comment (ahal) [inline explanations/best-practices/code-reviews.rst:7]: nit: What are

File: explanations/best-practices/code-reviews.rst
Comment (ahal) [inline explanations/best-practices/code-reviews.rst:16]: I think this belongs in the `Performing a review` section.

I suggest we also be a bit more explicit about the expectations:

> * Reviewers should endeavour to reply within 24hrs of the request, and are expected to reply within 48hrs.

We could also consider adding:

> If a review can't be performed within this time frame, the reviewer should redirect to someone who is able or at the very least leave a comment informing the author of the delay.

File: explanations/best-practices/code-reviews.rst
Comment (ahal) [inline explanations/best-practices/code-reviews.rst:17]: Suggestion:

> Using teams and properly scoped CODEOWNERS files can help with this.

File: explanations/best-practices/code-reviews.rst
Comment (ahal) [inline explanations/best-practices/code-reviews.rst:35]: My personal pet peeve is when a reviewer blocks waiting for more context but doesn't "Request changes". So I suggest adding something like:

> When requesting additional context, ensure you "Request changes". This puts the revision back in the author's queue and thereby helps avoid scenarios where they missed your question and the revision sits in limbo with both parties waiting for the other to respond.

File: explanations/best-practices/code-reviews.rst
Comment (ahal) [inline explanations/best-practices/code-reviews.rst:40]: My personal rule of thumb is to never request changes if I'm not comfortable with the patch landing as is (i.e, only use "r+ with these changes" if you wouldn't mind the author forgetting / missing the changes).

Though I'm not sure if we would want to put this into the doc, as there is a good argument to use "r+ with these changes" to help speed things up across time zones.

**Review by ahal — CHANGES_REQUESTED** [review]
Thanks, this is great! I have a few suggestions, and the `Timing` section should be moved, but otherwise this lgtm!

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:7]: Oops. Fixed!

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:16]: Okay. I moved it and tweaked your proposal notably to talk in business days instead of hours (to take weekends into account).

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:35]: I've never encountered a case where someone missed my review because I didn't hit "request changes". I know some people prefer just commenting on the PR in order to keep the conversation going. I'm not sure we need to make "request changes" a guideline for now. I'm happy to discuss this, though 🙂

Comment (ahal) [reply explanations/best-practices/code-reviews.rst:35]: It seems to happen on Phabricator more in my experience. And sure we can discuss more before including it.

**Review by ahal — APPROVED** [review]
Thanks for the fixes, lgtm!

File: explanations/best-practices/code-reviews.rst
Comment (jcristau) [inline explanations/best-practices/code-reviews.rst:16]: I would argue CODEOWNERS files do the opposite?
I think we also talked about requesting review from the group as a general rule, except in cases where review from a specific person as subject matter expert is desired?

File: explanations/best-practices/code-reviews.rst
Comment (jcristau) [inline explanations/best-practices/code-reviews.rst:22]: Not sure I understand the distinction that's being drawn here...  I'd argue if the bug has lots of back and forth, it's all the more important for the commit message(s) to have the information distilled down to one place.

**Review by jcristau — CHANGES_REQUESTED** [review]

File: explanations/best-practices/code-reviews.rst
Comment (ahal) [reply explanations/best-practices/code-reviews.rst:16]: Depends if the CODEOWNERS lists a single person or a group. In the latter case it can force people to use the group.. but good call out. It could also make things worse.

File: explanations/best-practices/code-reviews.rst
Comment (bhearsum) [reply explanations/best-practices/code-reviews.rst]: Maybe something changed after this comment, but I don't see anything relevant here?

File: explanations/best-practices/code-reviews.rst
Comment (bhearsum) [reply explanations/best-practices/code-reviews.rst:35]: Big +1 to this. Keeping the official "state" of the PR or revision in sync with actual expectations helps keep things running smoothly. (We should also make sure that everyone knows that "request changes" or similar is not a criticism or failing - just a reflection of where things are.)

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst]: Not sure what happened. The new sentence is: 

> New/changed tests must be read to ensure they check what the author intends.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:35]: Okay. I just added: 

> Use "Request Changes" to put the review back in the author's queue.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:16]: Okay. Should we mention anything about CODEOWNERS, then?

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:22]: Okay, I'll remove these points and leave it to the best judgement of the author. Happy to revisit this later if this needs clarifications.

**Review by bhearsum — APPROVED** [review]

**Review by gbrownmozilla — APPROVED** [review]

**Review by gabrielBusta — APPROVED** [review]

**Comment by hneiva** [comment]
image

😂

File: explanations/best-practices/code-reviews.rst
Comment (jcristau) [reply explanations/best-practices/code-reviews.rst:16]: I would leave it out.

File: explanations/best-practices/code-reviews.rst
Comment (jcristau) [inline explanations/best-practices/code-reviews.rst:17]: In my experience that doesn't actually happen in most cases, e.g. I don't remember seeing reminders there for mozilla-releng projects or for phabricator revisions.  Saying "will remind the reviewer" makes it sound a lot more definitive.

File: explanations/best-practices/code-reviews.rst
Comment (jcristau) [inline explanations/best-practices/code-reviews.rst:18]: s/Split your review/Split your PR/, or "Split your changes"?  It's not the review that's getting split.

**Review by jcristau — APPROVED** [review]
Some minor comments but nothing blocking for me anymore.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:18]: Fixed.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst:16]: Okay. Removed.

File: explanations/best-practices/code-reviews.rst
Comment (JohanLorenzo) [reply explanations/best-practices/code-reviews.rst]: Good point. Let's make this point more actionable then and allow the requester to remind the reviewer after 2 business days.

**Pull request merged**
